bin/merge.py

```diff
@@ -3,43 +3,140 @@
 
 import json
 from sys import argv
+from pathlib import Path
 
 
-def merge(filenames, filecount):
-    merged = {
-        "symbols": {},
-        "bigrams": {},
-        "trigrams": {},
-    }
-
-    # merge dictionaries
-    for filename in filenames:
-        with open(filename, "r") as corpus:
-            data = json.load(corpus)
-            for section in merged.keys():
-                for key, count in data[section].items():
-                    if key not in merged[section]:
-                        merged[section][key] = 0.0
-                    merged[section][key] += count / filecount
-
-    # sort the merged dictionary by symbol frequency (requires CPython 3.6+)
-    def sort_by_frequency(table, precision=2):
-        sorted_dict = {}
-        for key, count in sorted(table.items(), key=lambda x: -x[1]):
-            freq = round(count, precision)
-            if freq > 0:
-                sorted_dict[key] = freq
-        return sorted_dict
-
-    results = {}
-    results["corpus"] = ""
-    results["symbols"] = sort_by_frequency(merged["symbols"])
-    results["bigrams"] = sort_by_frequency(merged["bigrams"], 4)
-    results["trigrams"] = sort_by_frequency(merged["trigrams"])
-    return results
+# sort the merged dictionary by symbol frequency (requires CPython 3.6+)
+def _sort_ngram_by_frequency(table, precision=3):
+    sorted_dict = {}
+    for key, count in sorted(table.items(), key=lambda x: -x[1]):
+        freq = round(count, precision)
+        if freq > 0:
+            sorted_dict[key] = freq
+    return sorted_dict
+
+
+def sort_by_frequency(corpus: dict, precision=3):
+    for ngram in range(1, len(corpus["freq"].keys())):
+        ngram = str(ngram)
+        corpus["freq"][ngram] = _sort_ngram_by_frequency(
+            corpus["freq"][ngram], precision
+        )
+    return corpus
+
+
+def merge(filenames, output_name: str, mergetype: str, arg: dict = {}) -> dict:
+    """merge liste of JSON files output from `chardict` to get a sigle corpus file
+    the merge is on the number of character in each file"""
+
+    def _merge_dicts(corpus1: dict, corpus2: dict, mergetype: str, arg: dict) -> dict:
+        """concat two corpus dics into one based on number of characters"""
+        merge_dict = {
+            "name": "",
+            "freq": {},
+            "count": {},
+        }
+
+        # check corpus have the same number of n-grams for merge
+        if len(corpus1["freq"].keys()) != len(corpus2["freq"].keys()):
+            print(
+                f"Error: could not merge {corpus1["name"]} and {corpus2["name"]} :n-grams length is different ({ngram_length}-gram vs. {len(corpus["freq"].keys()+1)}-gram) ; skipping merge"
+            )
+            return corpus1
+
+        ngram_length = range(1, len(corpus1["freq"].keys()) + 1)
+        for n in ngram_length:
+            n = str(n)
```

Comment:
```suggestion
        for index in ngram_length:
            n = str(index)
```

It’s a good practice to avoid muting variables, i.e. variables should be constants as much as possible.

Muting a variable **and** changing its type is something Python allows (and I’ve done that a lot in JS as well, shame on me), but that’s something we should stop doing. Especially now that we use type annotations in our Python scripts.